From: Viswanath Kraleti <vkraleti@codeaurora.org>
Date: Thu, 26 Sep 2019 18:10:10 +0530
Subject: [PATCH] bitbake: Speed-up checksum calculations for .git directory

When a source directory is a git repository (contains .git) use
HEAD revision + git diff to calculate checksum instead of tracking
every file in .git folder.

---
 bitbake/lib/bb/checksum.py | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git bitbake/lib/bb/checksum.py bitbake/lib/bb/checksum.py
index 4e1598f..a5a18f9 100644
--- bitbake/lib/bb/checksum.py
+++ bitbake/lib/bb/checksum.py
@@ -96,6 +96,18 @@ class FileChecksumCache(MultiProcessCache):
             return checksum

         def checksum_dir(pth):
+            git_dir = os.path.join(pth, '.git')
+            if os.path.exists(git_dir):
+                import subprocess, hashlib
+                m = hashlib.md5()
+                head = subprocess.check_output("cd %s && git rev-parse HEAD" % pth, shell=True)
+                diff = subprocess.check_output("cd %s && git diff" % pth, shell=True)
+                m.update(head)
+                if diff:
+                    m.update(diff)
+
+                return [(pth, m.hexdigest())]
+
             # Handle directories recursively
             if pth == "/":
                 bb.fatal("Refusing to checksum /")
--
1.9.1
